keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+- Deny: Traffic that matches a `Deny` rule will be dropped. 
+- Allow: Traffic that matches an `Allow` rule will be allowed.
+
+AdminNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic with malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the `Deny` rules are deleted
+or modified by the admin.
+
+On the other hand, the `Allow` rules can be used to call out traffic in the cluster
+that needs to be allowed for certain components to work as expected (egress to
+CoreDNS for example). Those traffic should not be blocked when developers apply
+NetworkPolicy to their Namespaces which isolates the workloads.
+
+AdminNetworkPolicy `Pass` rules allows an admin to delegate security posture for
+certain traffic to the Namespace owners by overriding any lower priority Allow or 
+Deny rules. For example, allowing intra-tenant traffic can be delegated to tenant 
+admins explicitly by the cluster admin with the use of `Pass` rules.
+
+#### Priority
+
+The policy instances will be ordered based on the numeric priority assigned to the
+ANP. A larger number corresponds to a higher precedence. For example, ANP with `priority`
+set to "1" will be considered as the policy with the lowest precedence, while "999" will
+be considered as highest precedence. Any positive priority, however, will have
+higher precedence over the namespaced NetworkPolicy instances in the cluster.
+
+Additionally, the special priority "0" can be used in the priority field to indicate
+that the rules in that policy instance shall be created at a priority lower than the
+Namespaced NetworkPolicies. 
+
+The relative precedence of the rules within a ANP will be determined by the order in which the
+rule is written. Thus, a rule that appears at the top of the ingress/egress rules would take the
+highest precedence.
+
+Conflict resolution: Two policies are considered to be conflicting if they are assigned the
+same `priority` and apply to the same resources or a union of resources. In order to avoid such conflicts, 
+we propose to include tooling for ANP resources to help alert the admin to potentially ambiguous ANP
+priority scenarios, more details in [risks and mitigations](#risks-and-mitigations). However, ultimately
+it will be the job of the CNI to decide how to handle overlapping priority situations in the respective 
+implementations. 
+
+For conformance and portability purposes, it will be required that implementations
+support the range of 1000 numeric ANP priorities (0 through 999) per cluster and a maximum of 100 rules
+within an ANP instance. This ensures that policy manifests written using priorities
+0 through 999 are portable across clusters and implementations without the
+need to re-write policy manifests or re-adjust priority allocations and also ensures we maintain 
+safe upper bounds. 
+
+#### Multiple AppliedTos per policy instance
```

Comment:
I commented with a similar sentiment just below: https://github.com/kubernetes/enhancements/pull/2522/files#r791243604

Reply:
One instant rational for this (Aside from minimizing the  duplication of yaml + explosion in # of policies) is that this allows us to create dynamic level "0"  ANPs without having to supply multiple "baseline" priorities (i.e negative #s or something like that)

Reply:
Also I've seen thousands of duplicated netpol objects because we don't allow this abstraction

Reply:
> Also I've seen thousands of duplicated netpol objects because we don't allow this abstraction

Is this really a problem? Sure in some cases this might be a result of people using the API poorly, but in some cases maybe not. With this approach users also have a new option to use the API poorly by putting tons of unrelated rules in the same ANP.

> without having to supply multiple "baseline" priorities 

As discussed elsewhere, maybe we should have multiple baseline priorities.

Reply:
NetPol doesn't allow "same namespace"  and "same labels", though - can you show examples where dups would be needed with those?

Reply:
So off the top of my head, what if an admin wants to use a single priority level/ single ANP object to define all the tenants in a cluster for easier readability/ organization? This would allow them to do so, while if we didn't have multiple appliedTos we would need an individual CNPs for each tenant 

And ultimately the best part about it is the Admin can choose to do it either way....

Reply:
And the end of the day I keep going back to "do we really NOT need multiple targets?" and can't really come up with anything super concrete, although one of y'all might be able to

Reply:
> And ultimately the best part about it is the Admin can choose to do it either way....

This is the part I'm afraid of :laughing: 

> object to define all the tenants in a cluster for easier readability/ organization? 

What do you mean by "define" in this scenario? What sort of rules are in this policy? I'm still not fully understanding the use case. For "rubber-stamp" rules per-tenant, "same namespace" and "same labels" should do the trick with a single rule, right? For scenarios where the rules aren't the same per-tenant, I would think different ANP objects is the more clear organizational pattern to follow.

Reply:
This came up in a meeting as well, key takeaways: 
- For now we'll rename `AppliedTo` to `Subject` to further align with other APIs 
- Multiple Applied To's will be removed for alpha, and we'll re-analyze wether or not we need it for Beta